### PR TITLE
Make LogitsProcessor compatible with torch.compile

### LINKED ISSUE #28981: tracker: `generate` compatibility with `torch.compile`
# `generate` 🤜 🤛  `torch.compile`

Part of the [PyTorch 2024 H2 roadmap](https://drive.google.com/file/d/1Ucm17fyUeF0PWSd2g7jonM144XMx2t2r/view).

This issue is a tracker of the compatibility between `.generate` and `torch.compile` ([intro docs by pytorch](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html)). The goal is to enable `fullgraph=True` compilation on the main `generate` use cases.

⚠️ Is *your* `generate` use case not covered by this tracker? Check if it was requested below and upvote it if it was. Otherwise, add a comment. We will consider expanding the selection below on widely requested use cases 🤗 

### Decoding Strategies (end-to-end compilation)
- [x] `greedy_search` / `sample` are compatible (https://github.com/huggingface/transformers/pull/30788)
- [ ] `beam_search` / `beam_sample` are compatible, depends on the step above
- [ ] `assisted_decoding` (aka speculative decoding) is compatible, depends on the steps above

### Generate Flags and Options
- [ ] all `LogitsProcessor` classes were checked for compatibility (and the appropriate exceptions are raised when not compatible)
- [ ] all `StoppingCriteria` classes were checked for compatibility (and the appropriate exceptions are raised when not compatible)

### Models

Notes:
1. models tagged as "important models" in our CI + popular models
2. language models released starting from v4.42 should ALL support compile

Decoder-only:
- [x] GPT-J is compatible (#31421)
- [ ] GPT2 is compatible
- [x] Llama is compatible (#27931)
- [x] Gemma is compatible (#29167)
- [x] Llava is compatible (https://github.com/huggingface/transformers/pull/34502 removed dynamic control flow)
- [ ] Llava-Next is compatible (#29891)
- [x] Mistral is compatible (https://github.com/huggingface/transformers/pull/30642)
- [ ] Mixtral is compatible (https://github.com/huggingface/transformers/pull/30793/)
- [X] Phi is compatible (https://github.com/huggingface/transformers/pull/32617)
- [ ] Phi3 is compatible (https://github.com/huggingface/transformers/pull/30688)
- [X] BLOOM is compatible (note: this one might be tricky due to cache format) https://github.com/huggingface/transformers/pull/32617
- [x] Mamba is compatible (requested [here](https://github.com/huggingface/transformers/issues/29699#issuecomment-2072233223)) (PR: https://github.com/huggingface/transformers/pull/31247)
- [X] Persimmon is compatible (https://github.com/huggingface/transformers/pull/32617)
- [x] Qwen2 https://github.com/huggingface/transformers/pull/32617
- [x] Qwen2-VL https://github.com/huggingface/transformers/pull/32617
- [x] Falcon https://github.com/huggingface/transformers/pull/32617
- [x] GPTNeoX https://github.com/huggingface/transformers/pull/32617
- [x] Starcoder2 https://github.com/huggingface/transformers/pull/32617
- [x] StableLM https://github.com/huggingface/transformers/pull/32617

Encoder-decoder:
- [ ] BART is compatible (https://github.com/huggingface/transformers/pull/35314)
- [x] T5 is compatible (https://github.com/huggingface/transformers/pull/34089)
- [x] Whisper is compatible (https://github.com/huggingface/transformers/pull/31166)

### Quantization
- [ ] BNB support
- [ ] GPTQ support
- [ ] AWQ support

### Others
- [x] We have a benchmark script to quickly compare the impact of PRs
- [x] Add section to existing docs on the topic
- [x] Confirm that pipelines work after compiling generate

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Small part of the issue #28981 . This PR makes sure that Logits Processor and Stopping Criteria are compatible with `torch.compile` when `fullgraph=True`. The changes were tested with dummy inputs and logits and also with Llama. For now only the Processors used in `generate` were checked, those that are used in bark/whisper models can be checked later if needed.

The below processors are not compatible, exceptions will be added later:
- EncoderNoRepeatNGramLogitsProcessor and NoRepeatNGramLogitsProcessor -> tries to get a value from dict, which is input dependent
- PrefixConstrainedLogitsProcessor -> relies on user provided functions, which mostly probably are also input dependent
- SequenceBiasLogitsProcessor will not work at the same time with NoBadWordsProcessor, only one needs to be defined -> both call the same `_prepare_bias_variables`, which leads to recompiling it the second time we call with new arguments. Can be fixed if we either merge them into one processor or separate as two distinct.
- UnbatchedClassifierFreeGuidanceLogitsProcessor -> calls the model forward, current Llama with sdpa failed due to providing not None attention_mask.
- MaxTimeCriteria -> uses built-in time.time()

FYI @gante

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ class LogitsProcessor:
     """Abstract base class for all logit processors that can be applied during generation."""
 
     @add_start_docstrings(LOGITS_PROCESSOR_INPUTS_DOCSTRING)
-    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, cur_len: int) -> torch.FloatTensor:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:48]: Don't forget to update `LOGITS_PROCESSOR_INPUTS_DOCSTRING` accordingly (`LOGITS_PROCESSOR_INPUTS_DOCSTRING` is defined a few lines above)

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1411,12 +1417,11 @@ def __init__(self, bos_token_id: int):
...
-    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
-        cur_len = input_ids.shape[-1]
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, cur_len: int) -> torch.FloatTensor:
         if cur_len == 1:
-            num_tokens = scores.shape[1]
-            scores[:, [i for i in range(num_tokens) if i != self.bos_token_id]] = -float("inf")
-            scores[:, self.bos_token_id] = 0
+            mask = torch.full_like(scores, -math.inf)
+            mask[:, self.bos_token_id] = 0
+            scores = scores + mask
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1424]: Although technically correct (BOS will be selected), it is not fully backwards compatible: previously, `scores[:, self.bos_token_id]` was set to `0`, now it will not be set.

Perhaps `scores = mask`? Or the whole thing could also be something like `scores = torch.where(torch.arange(scores.shape[1]) == self.bos_token_id, 0, -math.inf))`

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -1460,13 +1465,11 @@ def __init__(self, max_length: int, eos_token_id: Union[int, List[int]]):
...
-        cur_len = input_ids.shape[-1]
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, cur_len: int) -> torch.FloatTensor:
         if cur_len == self.max_length - 1:
-            num_tokens = scores.shape[1]
-            scores[:, [i for i in range(num_tokens) if i not in self.eos_token_id]] = -float("inf")
-            for i in self.eos_token_id:
-                scores[:, i] = 0
+            mask = torch.full_like(scores, -math.inf)
+            mask[:, self.eos_token_id] = 0
+            scores = scores + mask
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1472]: (same as above)

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1480,13 +1483,17 @@ class InfNanRemoveLogitsProcessor(LogitsProcessor):
...
+        scores = scores.masked_scatter(scores != scores, zeros)
+
+        maxs = torch.ones_like(scores) * torch.finfo(scores.dtype).max
+        mins = torch.ones_like(scores) * torch.finfo(scores.dtype).min
 
         # set all +/-inf values to max/min possible value
-        scores[scores == float("inf")] = torch.finfo(scores.dtype).max
-        scores[scores == float("-inf")] = torch.finfo(scores.dtype).min
+        scores = scores.masked_scatter(scores == float("inf"), maxs)
+        scores = scores.masked_scatter(scores == -float("inf"), mins)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1496]: nit: we don't need to create `maxs`/`mins` with `scores = torch.where(scores == float("inf"), torch.finfo(scores.dtype).max, scores)`, which is (IMO) easier to read

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -1572,14 +1579,15 @@ def __init__(
         self.eos_token_id = eos_token_id
 
     @add_start_docstrings(LOGITS_PROCESSOR_INPUTS_DOCSTRING)
-    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
-        cur_len = input_ids.shape[-1]
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, cur_len: int) -> torch.FloatTensor:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1582]: Looking at this one, there is a chance we need to make `cur_len` a tensor, and not an `int`. If torch's compilation behaves anything like TF's, then integers will trigger a new compilation each time it is called with a different value 🤔 

Try to confirm that a recompilation is not happening as you call this one with different `cur_len`. If a recompilation is happening, try changing `cur_len` to a tensor. Avoiding recompilations is crucial for performance.

**Review by gante — COMMENTED** [review]
Looking good! I've added a few comments below

Note: the `fixes ...` in the PR header will close the tracker, which we don't want. In this case, avoid using one of the [magic keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) 🤗

File: src/transformers/generation/logits_process.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/generation/logits_process.py:1616]: Yes, you are right, `cur_len` as int causes recompilation every time the processor is called. But when we pass it in as a tensor, it will raise en error when trying to get tensor's value and compare the current length.

There is actually a way to use `functorch.cond` and then we can avoid graph breaks and recompilation, but it does not work with inductor backend right now. Pytorch devs said that it will be ready to use and efficient around the end of Q1.

**Comment by zucchini-nlp** [comment]
@gante I went through the comments and fixed where possible. I am wondering if it is a good idea to add warnings as I did? Maybe there is a better way to do it, so that the users do not see lots of unrelated warning. I guess not everyone will use `compile` to generate

**Comment by zucchini-nlp** [comment]
@gante Ready to review. I fixed tests and the generation utils to work with "cur_len", everything runs successfully in my machine.

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -34,6 +34,9 @@
         scores (`torch.FloatTensor` of shape `(batch_size, config.vocab_size)`):
             Prediction scores of a language modeling head. These can be logits for each vocabulary when not using beam
             search or log softmax for each vocabulary token when using beam search
+        cur_len: (`int`): The current sequence length of generated text. For compatibility with `torch.compile`, `input_ids`
+            sequence length is the maxiumum length that can be generated, and `cur_len` indicates the length that was actually
```
Comment (gante) [inline src/transformers/generation/logits_process.py:38]:
```suggestion
            sequence length is the maximum length that can be generated, and `cur_len` indicates the length that was actually
```

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -78,6 +81,9 @@ def __call__(
             scores (`torch.FloatTensor` of shape `(batch_size, config.vocab_size)`):
                 Prediction scores of a language modeling head. These can be logits for each vocabulary when not using
                 beam search or log softmax for each vocabulary token when using beam search
+            cur_len: (`int`): The current sequence length of generated text. For compatibility with `torch.compile`, `input_ids`
+                sequence length is the maxiumum length that can be generated, and `cur_len` indicates the length that was actually
```
Comment (gante) [inline src/transformers/generation/logits_process.py:85]:
```suggestion
                sequence length is the maximum length that can be generated, and `cur_len` indicates the length that was actually
```

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -146,6 +152,10 @@ def __init__(self, min_length: int, eos_token_id: Union[int, List[int]]):
             eos_token_id = [eos_token_id]
         if not all(isinstance(i, int) for i in eos_token_id) or any(i < 0 for i in eos_token_id):
             logger.warning(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
+        logger.warning(
+            f"{self.__class__.__name__} will cause recompilations for every new token if used with `torch.compile`."
+            "do not add `min_length` in generation config for efficient compilation"
```
Comment (gante) [inline src/transformers/generation/logits_process.py:157]:
```suggestion
            "Do not add `min_length` in generation config for efficient compilation"
```
(this change should be propagated to the other warnings)

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -863,6 +877,10 @@ class NoRepeatNGramLogitsProcessor(LogitsProcessor):
     def __init__(self, ngram_size: int):
         if not isinstance(ngram_size, int) or ngram_size <= 0:
             raise ValueError(f"`ngram_size` has to be a strictly positive integer, but is {ngram_size}")
+        logger.warning(
```
Comment (gante) [inline src/transformers/generation/logits_process.py:880]: These cases that prevent compilation should be exceptions instead. More precisely, if possible, the original exception should be caught, the message you wrote here gets appended at the end, and then the same exception is reraised.

That way, the user has a full stack trace and, if for some reason torch fixes the issue on their end, the problem becomes automatically solved on our end as well!

**Review by gante — COMMENTED** [review]
While re-reviewing the PR, I noticed that we were not taking backward compatibility (BC) in mind 😬 The logits processors are public classes, and thus we need to warn users of the new `cur_len` argument. We should:
1. set `cur_len=None` in ALL signatures.
2. when `cur_len=None`, set it to `input_ids.shape[1]` (1. + 2. make these changes BC)
3. add a deprecation cycle for the case where `cur_len is None` in ALL logits processors. A deprecation cycle looks like [this](https://github.com/huggingface/transformers/blob/1d0ea7abe0a3d73414545fd401742ad25b0e7795/src/transformers/generation_utils.py#L24) (in this case, with a `warning_once`, see below). Since this is a substantial change, let's give the users 4 minor versions to adapt (i.e. target to remove in v4.43).

Other warning-related comments:
1. you should be able to emit the warning using [this function](https://github.com/huggingface/transformers/blob/1d0ea7abe0a3d73414545fd401742ad25b0e7795/src/transformers/utils/import_utils.py#L529), as in [this example](https://github.com/huggingface/transformers/blob/1d0ea7abe0a3d73414545fd401742ad25b0e7795/src/transformers/modeling_utils.py#L4162).
2. replace `logger.warning` with `logger.warning_once`. Otherwise there is a warning at each generation step :)

File: src/transformers/generation/logits_process.py (hunk 9 above)
Comment (zucchini-nlp) [reply src/transformers/generation/logits_process.py]: I found out that loggings/prints/warning cause a graph break in the current version of pytorch. There is an [open PR](https://github.com/pytorch/pytorch/issues/93739) to fix it. 

I am committing a temporary change with all the logging and warning right now. Everything has to be tested when the pytorch PR gets merged. Current changes will not compile with fullgraph.
